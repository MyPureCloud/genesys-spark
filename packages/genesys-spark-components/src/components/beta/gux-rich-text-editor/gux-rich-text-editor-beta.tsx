import {
  Component,
  h,
  JSX,
  Element,
  Prop,
  readTask,
  State,
  Event,
  EventEmitter
} from '@stencil/core';
import { OnResize } from '@utils/decorator/on-resize';
import { hasSlot } from '@utils/dom/has-slot';
import { trackComponent } from '@utils/tracking/usage';
import { getActionsFromGroup } from './gux-rich-text-editor.service';
import { buildI18nForComponent, GetI18nValue } from 'i18n';
import translationResources from './gux-rich-text-editor-action/i18n/en.json';
import { afterNextRenderTimeout } from '@utils/dom/after-next-render';

/**
 * @slot typographical-emphasis - Slot for typographical actions.
 * @slot text-styling - Slot for text-styling actions.
 * @slot lists-indentation - Slot for lists and indentation actions.
 * @slot inserting - Slot for inserting actions.
 * @slot global-action - Slot for global action.
 * @slot editor - Slot for the editor.
 */

@Component({
  tag: 'gux-rich-text-editor-beta',
  styleUrl: 'gux-rich-text-editor.scss',
  shadow: true
})
export class GuxRichTextEditor {
  private i18n: GetI18nValue;

  @Element()
  root: HTMLElement;

  @Prop()
  disabled: boolean = false;

  @State()
  typographicalEmphasisActions: string[] = [];

  @State()
  textStylingActions: string[] = [];

  @State()
  listsAndIndentationActions: string[] = [];

  @State()
  insertingActions: string[] = [];

  // This event is emitted when an action has been selected from the menu in the shadowDOM.
  @Event({ composed: true })
  guxToggleAction: EventEmitter<string>;

  @OnResize()
  checkResponsiveLayout(): void {
    readTask(() => {
      this.handleOverflow();
      this.getHiddenActions();
    });
  }

  async componentWillLoad(): Promise<void> {
    trackComponent(this.root);
    this.i18n = await buildI18nForComponent(this.root, translationResources);
  }

  componentDidLoad(): void {
    // This timeout is required to calculate the correct size of the containers when the component loads. By including a timeout of 1 second the containers calculate correctly.
    afterNextRenderTimeout(() => {
      this.checkResponsiveLayout();
    });
  }

  componentDidUpdate(): void {
    this.handleOverflow();
  }

  private isOverFlowing(): boolean {
    const container = this.root?.shadowRoot.querySelector(
      '.gux-rich-text-editor-toolbar-container'
    );
    // A group of gux-rich-text-editor-action-group.
    const children = Array.from(container.children);
    // Calculate the total width of all children excluding hidden (gux-rich-text-editor-action-groups).
    const totalWidth = children
      .filter(child => !child.classList.contains('gux-hidden'))
      .reduce((sum, child) => sum + child.clientWidth, 0);

    return totalWidth > container.clientWidth;
  }

  private handleOverflow(): void {
    const actionGroups = Array.from(
      this.root.querySelectorAll('gux-rich-text-editor-action-group')
    );

    // Hide last visible action group until there's no overflow.
    while (this.isOverFlowing()) {
      let hidden = false;
      for (let i = actionGroups.length - 1; i >= 0; i--) {
        if (!actionGroups[i].classList.contains('gux-hidden')) {
          actionGroups[i].classList.add('gux-hidden');
          hidden = true;
          break;
        }
      }
      // If no more action groups to hide, exit loop
      if (!hidden) {
        break;
      }
    }

    // Show previously hidden groups if space is available
    for (const group of actionGroups) {
      if (group.classList.contains('gux-hidden')) {
        group.classList.remove('gux-hidden');

        // If it causes overflow again, re-hide the group and stop the loop
        if (this.isOverFlowing()) {
          group.classList.add('gux-hidden');
          break;
        }
      }
    }
  }
  private getHiddenActions(): void {
    this.typographicalEmphasisActions = getActionsFromGroup(
      this.root,
      'gux-rich-text-editor-action-group[slot="typographical-emphasis"]',
      'gux-hidden'
    );

    this.listsAndIndentationActions = getActionsFromGroup(
      this.root,
      'gux-rich-text-editor-action-group[slot="lists-indentation"]',
      'gux-hidden'
    );

    this.textStylingActions = getActionsFromGroup(
      this.root,
      'gux-rich-text-editor-action-group[slot="text-styling"]',
      'gux-hidden'
    );

    this.insertingActions = getActionsFromGroup(
      this.root,
      'gux-rich-text-editor-action-group[slot="inserting"]',
      'gux-hidden'
    );
  }

  private getAllActions(): string[] {
    return [
      ...this.textStylingActions,
      ...this.typographicalEmphasisActions,
      ...this.listsAndIndentationActions,
      ...this.insertingActions
    ];
  }

  private categorizeActions(allActions: string[]) {
    const richStylePrefix = 'rich-style-';
    const highlightPrefix = 'highlight-';
    const richStyleActions: string[] = [];
    const highlightActions: string[] = [];
    const filteredActions: string[] = [];

    allActions.forEach(action => {
      if (action.startsWith(richStylePrefix)) {
        richStyleActions.push(action.slice(richStylePrefix.length));
      } else if (action.startsWith(highlightPrefix)) {
        highlightActions.push(action.slice(highlightPrefix.length));
      } else {
        filteredActions.push(action);
      }
    });

    return { richStyleActions, highlightActions, filteredActions };
  }

  private renderListItems(actions: string[], useInnerHTML = false) {
    return actions.map((action, index) => {
      const actionText = action.replace(/<\/?[^>]+(>|$)/g, ''); // Remove HTML tags

      return (
        <gux-rich-style-list-item
          onClick={() => this.guxToggleAction.emit(actionText)}
          key={index}
          value={actionText}
          {...(useInnerHTML ? { innerHTML: action } : {})}
        >
          {!useInnerHTML && (this.i18n(actionText) || actionText)}
        </gux-rich-style-list-item>
      );
    });
  }

  private renderSubList(
    labelKey: string,
    actions: string[],
    useInnerHTML = false
  ) {
    if (actions.length > 0) {
      return (
        <gux-rich-text-editor-sub-list label={this.i18n(labelKey)}>
          {this.renderListItems(actions, useInnerHTML)}
        </gux-rich-text-editor-sub-list>
      );
    }
  }

  private renderHighlightSubList(actions: string[]) {
    if (actions.length > 0) {
      return (
        <gux-rich-text-editor-sub-list label={this.i18n('textHighlight')}>
          {this.renderListItems(actions)}
          <gux-rich-style-list-item
            onClick={() => this.guxToggleAction.emit('noHighlightAction')}
            value="noHighlight"
          >
            <gux-truncate max-lines={1}>
              {this.i18n('noHighlight')}
            </gux-truncate>
          </gux-rich-style-list-item>
        </gux-rich-text-editor-sub-list>
      );
    }
  }

  private renderTextEditorMenu(): JSX.Element | null {
    const allActions = this.getAllActions();
    if (allActions.length > 0) {
      const { richStyleActions, highlightActions, filteredActions } =
        this.categorizeActions(allActions);

      return (
        <gux-rich-text-editor-menu>
          {this.renderListItems(filteredActions)}
          {this.renderSubList('richStyle', richStyleActions, true)}
          {this.renderHighlightSubList(highlightActions)}
        </gux-rich-text-editor-menu>
      );
    }
  }

  private renderSlot(
    slotName: string,
    containerClass: string
  ): JSX.Element | null {
    if (hasSlot(this.root, slotName)) {
      return (
        <div class={containerClass}>
          <slot name={slotName}></slot>
        </div>
      ) as JSX.Element;
    }
    return null;
  }

  private renderTypographicalEmphasis(): JSX.Element | null {
    return this.renderSlot(
      'typographical-emphasis',
      'gux-typographical-emphasis-container'
    );
  }

  private renderTextStyling(): JSX.Element | null {
    return this.renderSlot('text-styling', 'gux-text-styling-container');
  }

  private renderListsIndentation(): JSX.Element | null {
    return this.renderSlot(
      'lists-indentation',
      'gux-lists-indentation-container'
    );
  }

  private renderInserting(): JSX.Element | null {
    return this.renderSlot('inserting', 'gux-inserting-container');
  }

  private renderGlobalAction(): JSX.Element | null {
    return this.renderSlot('global-action', 'gux-global-action-container');
  }

  render(): JSX.Element {
    return (
      <div
        class={{
          'gux-rich-text-editor-container': true,
          'gux-disabled': this.disabled
        }}
      >
        <div class="gux-rich-text-editor-toolbar-container">
          {this.renderTypographicalEmphasis()}
          {this.renderTextStyling()}
          {this.renderListsIndentation()}
          {this.renderInserting()}
          {this.renderTextEditorMenu()}
          {this.renderGlobalAction()}
        </div>
        <slot name="editor"></slot>
      </div>
    ) as JSX.Element;
  }
}
