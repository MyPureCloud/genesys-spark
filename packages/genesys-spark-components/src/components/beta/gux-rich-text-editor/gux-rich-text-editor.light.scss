@use '~genesys-spark/dist/beta/tokens/flare/ui.scss';

.editor-styles {
  font-family: ui.$gse-ui-rte-paragraph-md-regular-fontFamily;
  font-size: ui.$gse-ui-rte-paragraph-md-regular-fontSize;
  font-weight: ui.$gse-ui-rte-paragraph-md-regular-fontWeight;
  line-height: ui.$gse-ui-rte-paragraph-md-regular-lineHeight;
  color: ui.$gse-ui-rte-paragraph-foregroundColor;
  outline: none;

  p {
    margin: 0;
  }

  a {
    cursor: pointer;
  }

  blockquote {
    margin: ui.$gse-ui-rte-quoteBlock-margin;
    border-inline-start: ui.$gse-ui-rte-quoteBlock-borderLeft-width solid
      ui.$gse-ui-rte-quoteBlock-borderLeft-color;

    p {
      margin-inline-start: ui.$gse-ui-rte-quoteBlock-container-gap;
      font-family: ui.$gse-ui-rte-quoteBlock-md-regular-fontFamily;
      font-size: ui.$gse-ui-rte-quoteBlock-md-regular-fontSize;
      font-weight: ui.$gse-ui-rte-quoteBlock-md-regular-fontWeight;
      line-height: ui.$gse-ui-rte-quoteBlock-md-regular-lineHeight;
      color: ui.$gse-ui-rte-quoteBlock-foregroundColor;
    }
  }

  pre {
    padding: ui.$gse-ui-rte-codeBlock-padding;
    background-color: ui.$gse-ui-rte-codeBlock-backgroundColor;
    border: ui.$gse-ui-rte-codeBlock-border-width
      ui.$gse-ui-rte-codeBlock-border-style
      ui.$gse-ui-rte-codeBlock-border-color;
    border-radius: ui.$gse-ui-rte-codeBlock-borderRadius;

    code {
      font-family: ui.$gse-ui-rte-codeBlock-sm-regular-fontFamily;
      font-size: ui.$gse-ui-rte-codeBlock-sm-regular-fontSize;
      font-weight: ui.$gse-ui-rte-codeBlock-sm-regular-fontWeight;
      line-height: ui.$gse-ui-rte-codeBlock-sm-regular-lineHeight;
      color: ui.$gse-ui-rte-codeBlock-foregroundColor;
      white-space: pre-wrap;
    }
  }
}
