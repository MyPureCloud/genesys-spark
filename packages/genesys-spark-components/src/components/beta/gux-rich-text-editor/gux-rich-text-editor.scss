@use '~genesys-spark/dist/beta/tokens/flare/ui.scss';

:host {
  display: block;
}

:host([style*='resize']) {
  overflow: auto;
  resize: inherit;
}

.gux-rich-text-editor-container {
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  gap: ui.$gse-ui-rte-innerContainer-gap;
  inline-size: 100%;
  max-inline-size: 100%;
  block-size: 100%;
  max-block-size: 100%;
  padding: ui.$gse-ui-rte-padding;
  overflow: hidden scroll;
  background-color: ui.$gse-ui-formControl-input-backgroundColor;
  border: ui.$gse-ui-rte-mainContainer-default-border-width
    ui.$gse-ui-rte-mainContainer-default-border-color
    ui.$gse-ui-rte-mainContainer-default-border-style;
  border-radius: ui.$gse-ui-rte-container-borderRadius;

  &.gux-toolbar-hidden {
    gap: 0;
  }

  .gux-rich-text-editor-toolbar-container {
    display: flex;
    flex-direction: row;
    justify-content: flex-start;

    .gux-global-action-container {
      margin-inline-start: auto;
    }
  }

  &.gux-disabled {
    pointer-events: none;
    cursor: default;
    user-select: none;
    border: ui.$gse-ui-rte-mainContainer-disabled-border-width
      ui.$gse-ui-rte-mainContainer-disabled-border-color
      ui.$gse-ui-rte-mainContainer-disabled-border-style;
    opacity: 0.5;
  }

  &:active:enabled {
    border: ui.$gse-ui-rte-mainContainer-active-border-width
      ui.$gse-ui-rte-mainContainer-active-border-color
      ui.$gse-ui-rte-mainContainer-active-border-style;
  }

  &:focus-within:not(.gux-disabled),
  &:hover:not(.gux-disabled) {
    border: ui.$gse-ui-rte-mainContainer-hover-border-width
      ui.$gse-ui-rte-mainContainer-hover-border-color
      ui.$gse-ui-rte-mainContainer-hover-border-style;
  }

  &:focus-within:not(.gux-disabled) {
    outline: ui.$gse-ui-rte-mainContainer-focus-border-width
      ui.$gse-ui-rte-mainContainer-focus-border-color
      ui.$gse-ui-rte-mainContainer-focus-border-style;
  }

  .gux-editor-container {
    position: relative;
    flex: 1;
  }

  .gux-drop-overlay {
    position: absolute;
    inset: 0;
    z-index: 1000;
    display: flex;
    align-items: center;
    justify-content: center;
    font-size: 18px;
    color: #666;
    pointer-events: none;
    background: rgb(255 255 255);
    border: 1px dashed black;
  }

  .gux-files-area {
    display: flex;
    flex-direction: row;
    gap: 8px;
    padding: 16px;
    border-block-start: 1px solid #e0e0e0;
  }

  .gux-file-item {
    display: flex;
    align-items: center;
    justify-content: space-between;
    inline-size: 150px;
    padding-block: 8px;
    padding-inline: 12px;
    font-size: 14px;
    color: #333;
    background: #f5f6fa;
    border-radius: 8px;
    box-shadow: 0 2px 4px rgb(0 0 0 / 10%);
  }
}
