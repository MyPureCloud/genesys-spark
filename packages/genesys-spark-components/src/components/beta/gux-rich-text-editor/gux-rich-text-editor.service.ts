import { getClosestElement } from 'genesys-spark-utils/get-closest-element';
import { GuxRichTextEditorActionTypes } from './gux-rich-text-editor-action/gux-rich-text-editor-action.types';

export function hasDisabledParent(root: HTMLElement): boolean {
  const getParent = getClosestElement(
    root,
    'gux-rich-text-editor-beta'
  ) as HTMLGuxRichTextEditorBetaElement;

  return getParent?.disabled;
}

export function returnActionTypeIcon(
  action: GuxRichTextEditorActionTypes
): string {
  switch (action) {
    case 'bold':
      return 'fa/bold-regular';
    case 'italic':
      return 'fa/italic-regular';
    case 'underline':
      return 'fa/underline-regular';
    case 'strike':
      return 'fa/strikethrough-regular';
    case 'codeblock':
      return 'fa/code-regular';
    case 'blockQuote':
      return 'fa/quote-right-regular';
    case 'orderedList':
      return 'fa/list-ol-regular';
    case 'bulletList':
      return 'fa/list-ul-regular';
    case 'clearFormatting':
      return 'fa/eraser-regular';
    case 'undo':
      return 'custom/undo-regular';
    case 'redo':
      return 'custom/redo-regular';
    case 'delete':
      return 'fa/trash-regular';
    default:
      return 'fa/bold-regular';
  }
}

/**
 * Extracts an attribute from a action and adds it to the actions array.
 */
function addAction(actions: string[], element: Element, attribute: string) {
  const value = (element as HTMLElement).getAttribute(attribute);
  if (value) {
    actions.push(value);
  }
}

/**
 * This function is used for lists eg rich-style and highlight.
 * Extracts a list of values from child elements and adds them to the actions array with a prefix. eg(rich-style and highlight).
 */
function addListActions(
  actions: string[],
  parent: Element,
  selector: string,
  prefix: string,
  attribute: string = ''
) {
  Array.from(parent.querySelectorAll(selector)).forEach(item => {
    const value = attribute
      ? (item as HTMLElement).getAttribute(attribute)
      : (item as HTMLElement).innerHTML;
    if (value) {
      actions.push(`${prefix}-${value}`);
    }
  });
}

export function getActionsFromGroup(
  root: HTMLElement,
  selector: string,
  hiddenClass: string
): string[] {
  const actionGroup = root.querySelector(selector);
  if (!actionGroup || !actionGroup.classList.contains(hiddenClass)) {
    return [];
  }

  return Array.from(actionGroup.children).reduce<string[]>((actions, child) => {
    const tagName = child.tagName;

    if (tagName === 'GUX-RICH-TEXT-EDITOR-ACTION') {
      addAction(actions, child, 'action');
    } else if (tagName === 'GUX-RICH-TEXT-EDITOR-ACTION-RICH-STYLE') {
      addListActions(actions, child, 'gux-rich-style-list-item', 'rich-style');
    } else if (tagName === 'GUX-RICH-TEXT-EDITOR-ACTION-TEXT-HIGHLIGHT') {
      addListActions(
        actions,
        child,
        'gux-rich-highlight-list-item',
        'highlight',
        'highlight'
      );
    }
    return actions;
  }, []);
}
