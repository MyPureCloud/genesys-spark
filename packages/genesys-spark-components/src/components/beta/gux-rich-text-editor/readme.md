# gux-rich-text-editor-beta



<!-- Auto Generated Below -->


## Properties

| Property   | Attribute  | Description | Type      | Default |
| ---------- | ---------- | ----------- | --------- | ------- |
| `disabled` | `disabled` |             | `boolean` | `false` |


## Slots

| Slot                       | Description                             |
| -------------------------- | --------------------------------------- |
| `"editor"`                 | Slot for the editor.                    |
| `"global-action"`          | Slot for global action.                 |
| `"inserting"`              | Slot for inserting actions.             |
| `"lists-indentation"`      | Slot for lists and indentation actions. |
| `"text-styling"`           | Slot for text-styling actions.          |
| `"typographical-emphasis"` | Slot for typographical actions.         |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
