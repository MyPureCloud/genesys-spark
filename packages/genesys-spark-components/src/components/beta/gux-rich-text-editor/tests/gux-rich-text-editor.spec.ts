import { checkRenders } from '@test/specTestUtils';
import { GuxRichTextEditor } from '../gux-rich-text-editor-beta';
import { renderConfigs } from './gux-rich-text-editor.common';

const components = [GuxRichTextEditor];

describe('gux-rich-text-editor', () => {
  describe('#render', () => {
    checkRenders(renderConfigs, components);
  });
});
