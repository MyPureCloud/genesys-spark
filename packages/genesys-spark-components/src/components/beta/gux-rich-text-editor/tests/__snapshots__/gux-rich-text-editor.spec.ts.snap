// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-rich-text-editor #render should render component as expected (1) 1`] = `
<gux-rich-text-editor-beta>
  <template shadowrootmode="open">
    <div class="gux-rich-text-editor-container">
      <div class="gux-rich-text-editor-toolbar-container">
        <div class="gux-typographical-emphasis-container">
          <slot name="typographical-emphasis"></slot>
        </div>
        <div class="gux-text-styling-container">
          <slot name="text-styling"></slot>
        </div>
        <div class="gux-lists-indentation-container">
          <slot name="lists-indentation"></slot>
        </div>
        <div class="gux-inserting-container">
          <slot name="inserting"></slot>
        </div>
        <div class="gux-global-action-container">
          <slot name="global-action"></slot>
        </div>
      </div>
      <slot name="editor"></slot>
    </div>
  </template>
  <gux-rich-text-editor-action-group slot="typographical-emphasis">
    <gux-rich-text-editor-action action="bold"></gux-rich-text-editor-action>
    <gux-rich-text-editor-action action="italic"></gux-rich-text-editor-action>
    <gux-rich-text-editor-action action="underline"></gux-rich-text-editor-action>
    <gux-rich-text-editor-action action="strike"></gux-rich-text-editor-action>
  </gux-rich-text-editor-action-group>
  <gux-rich-text-editor-action-group slot="text-styling">
    <gux-rich-text-editor-action action="clearFormatting"></gux-rich-text-editor-action>
  </gux-rich-text-editor-action-group>
  <gux-rich-text-editor-action-group slot="lists-indentation">
    <gux-rich-text-editor-action action="bulletList"></gux-rich-text-editor-action>
    <gux-rich-text-editor-action action="orderedList"></gux-rich-text-editor-action>
  </gux-rich-text-editor-action-group>
  <gux-rich-text-editor-action-group slot="inserting">
    <gux-rich-text-editor-action action="codeblock"></gux-rich-text-editor-action>
    <gux-rich-text-editor-action action="blockQuote"></gux-rich-text-editor-action>
  </gux-rich-text-editor-action-group>
  <gux-rich-text-editor-action action="delete" slot="global-action"></gux-rich-text-editor-action>
  <div class="editorElement" slot="editor"></div>
</gux-rich-text-editor-beta>
`;

exports[`gux-rich-text-editor #render should render component as expected (2) 1`] = `
<gux-rich-text-editor-beta disabled="">
  <template shadowrootmode="open">
    <div class="gux-disabled gux-rich-text-editor-container">
      <div class="gux-rich-text-editor-toolbar-container">
        <div class="gux-typographical-emphasis-container">
          <slot name="typographical-emphasis"></slot>
        </div>
        <div class="gux-text-styling-container">
          <slot name="text-styling"></slot>
        </div>
        <div class="gux-lists-indentation-container">
          <slot name="lists-indentation"></slot>
        </div>
        <div class="gux-inserting-container">
          <slot name="inserting"></slot>
        </div>
        <div class="gux-global-action-container">
          <slot name="global-action"></slot>
        </div>
      </div>
      <slot name="editor"></slot>
    </div>
  </template>
  <gux-rich-text-editor-action-group slot="typographical-emphasis">
    <gux-rich-text-editor-action action="bold"></gux-rich-text-editor-action>
    <gux-rich-text-editor-action action="italic"></gux-rich-text-editor-action>
    <gux-rich-text-editor-action action="underline"></gux-rich-text-editor-action>
    <gux-rich-text-editor-action action="strike"></gux-rich-text-editor-action>
  </gux-rich-text-editor-action-group>
  <gux-rich-text-editor-action-group slot="text-styling">
    <gux-rich-text-editor-action action="clearFormatting"></gux-rich-text-editor-action>
  </gux-rich-text-editor-action-group>
  <gux-rich-text-editor-action-group slot="lists-indentation">
    <gux-rich-text-editor-action action="bulletList"></gux-rich-text-editor-action>
    <gux-rich-text-editor-action action="orderedList"></gux-rich-text-editor-action>
  </gux-rich-text-editor-action-group>
  <gux-rich-text-editor-action-group slot="inserting">
    <gux-rich-text-editor-action action="codeblock"></gux-rich-text-editor-action>
    <gux-rich-text-editor-action action="blockQuote"></gux-rich-text-editor-action>
  </gux-rich-text-editor-action-group>
  <gux-rich-text-editor-action action="delete" slot="global-action"></gux-rich-text-editor-action>
  <div class="editorElement" slot="editor"></div>
</gux-rich-text-editor-beta>
`;
